analyzer/src/main/kotlin/managers/GoMod.kt

```diff
@@ -235,54 +227,57 @@ class GoMod(
     }
 }
 
-private data class Edge(
-    val source: Identifier,
-    val target: Identifier
-)
-
-private fun Collection<Edge>.getNodes(): Set<Identifier> = flatMap { listOf(it.source, it.target) }.toSet()
-
-private fun Collection<Edge>.toPackageReferenceForest(
-    ignoreNode: (Identifier) -> Boolean = { false }
-): SortedSet<PackageReference> {
-    data class Node(
-        val id: Identifier,
-        val outgoingEdges: MutableSet<Identifier> = mutableSetOf(),
-        val incomingEdges: MutableSet<Identifier> = mutableSetOf()
-    )
+/**
+ * A type alias for a representation of the dependency graph used by this class. This is basically an adjacency list
+ * implemented by a map whose keys are package identifiers and whose values are the identifiers of packages these
+ * packages depend on.
+ */
+private typealias Graph = MutableMap<Identifier, Set<Identifier>>
 
-    val nodes = mutableMapOf<Identifier, Node>()
-    fun addNode(id: Identifier): Node? = if (!ignoreNode(id)) nodes.getOrPut(id) { Node(id) } else null
+/**
+ * Add an edge (i.e. a dependency relation) from [source] to [target] to this dependency graph. Add missing nodes if
+ * necessary.
+ */
+private fun Graph.addEdge(source: Identifier, target: Identifier) {
+    merge(source, setOf(target)) { set, _ -> set + target }
+    getOrPut(target) { emptySet() }
+}
 
-    forEach { edge ->
-        val source = addNode(edge.source)
-        val target = addNode(edge.target)
+/**
+ * Return a new [Graph] that is restricted to the given [nodes]. This can be used to construct graphs for specific
+ * scopes.
+ */
+private fun Graph.retainNodes(nodes: Set<Identifier>): Graph =
```

Comment:
I believe this is commonly called `subgraph`. So how about naming the function `subgraph()` or `getSubgraph()` ?